### PR TITLE
[BugFix] check table existence before heavy creating operation

### LINKED ISSUE #30388: "create table if not exist" shall check the table existence first
### Steps to reproduce the behavior (Required)

1. Create database
```
CREATE DATABASE demo;
```
2. Create table
```
use demo;
CREATE TABLE IF NOT EXISTS `sample` (
  `ds` date NULL COMMENT "",
  `id` int(11) NULL DEFAULT "0" COMMENT ""
) ENGINE=OLAP 
UNIQUE KEY(`ds`, `id`)
COMMENT "OLAP"
PARTITION BY RANGE(`ds`)
(
  PARTITION p20190101 VALUES [('2019-01-01'), ('2019-01-02'))
)
DISTRIBUTED BY HASH(`ds`, `id`) BUCKETS 8 
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "true",
"dynamic_partition.time_unit" = "DAY",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-2147483648",
"dynamic_partition.end" = "3",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "8",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);

```
The execution will be finished soon

3. Create the table if not exist with a large amount of partitions (the creation will be costly)

```
CREATE TABLE IF NOT EXISTS `sample` (
  `ds` date NULL COMMENT "",
  `id` int(11) NULL DEFAULT "0" COMMENT ""
) ENGINE=OLAP 
UNIQUE KEY(`ds`, `id`)
COMMENT "OLAP"
PARTITION BY RANGE(`ds`)
(
  PARTITION p20190101 VALUES [('2019-01-01'), ('2019-01-02')),
  ...
  PARTITION p20230907 VALUES [('2023-09-07'), ('2023-09-08'))
)
DISTRIBUTED BY HASH(`ds`, `id`) BUCKETS 8 
PROPERTIES (
"replication_num" = "1",
"dynamic_partition.enable" = "true",
"dynamic_partition.time_unit" = "DAY",
"dynamic_partition.time_zone" = "Asia/Shanghai",
"dynamic_partition.start" = "-2147483648",
"dynamic_partition.end" = "3",
"dynamic_partition.prefix" = "p",
"dynamic_partition.buckets" = "8",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);

Query OK, 0 rows affected (1 min 3.70 sec)
```
It will take a long time to complete and then give up new table creation eventually because of the table already exists. This phenomenon is outstanding especially in shared-data mode.

### Expected behavior (Required)

Fail Fast if table already exists

### Real behavior (Required)

Did everything necessary and give up at the end due to existence check.

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by kevincai**
* add the ahead existence check back which is removed in #25741
* Fixes #30388

## What type of PR is this:

- [X] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [X] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [X] I have checked the version labels which the pr will be auto-backported to the target branch
  - [X] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 4 / 8 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|8|50.00%|[797, 798, 800, 802]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 4 / 8 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|8|50.00%|[797, 798, 800, 802]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 4 / 8 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|4|8|50.00%|[797, 798, 800, 802]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 7 / 8 (87.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|8|87.50%|[800]|

**Review by gengjun-git — DISMISSED** [review]

**Review by wyb — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 1]
```diff
@@ -788,6 +788,23 @@ public boolean createTable(CreateTableStmt stmt) throws DdlException {
...
+        db.readLock();
+        try {
+            String tableName = stmt.getTableName();
+            if (db.getTable(tableName) != null) {
+                if (stmt.isSetIfNotExists()) {
+                    LOG.info("create table[{}] which already exists", tableName);
+                } else {
+                    ErrorReport.reportDdlException(ErrorCode.ERR_TABLE_EXISTS_ERROR, tableName);
+                }
+                return false;
```
[the remark this answers is not in the document]
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:802]: done

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 7 / 8 (87.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|8|87.50%|[798]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by HangyuanLiu — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**